// supermon/dist/lib/EventBus/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WatchEvents = exports.ModulesEvents = exports.ChildEvents = exports.ProcessEvents = void 0;
const events_1 = __importDefault(require("events"));
const logger_1 = __importDefault(require("../logger/logger"));
var ProcessEvents;
(function (ProcessEvents) {
    ProcessEvents["Start"] = "PROCESS_START";
})(ProcessEvents = exports.ProcessEvents || (exports.ProcessEvents = {}));
var ChildEvents;
(function (ChildEvents) {
    ChildEvents["Start"] = "CHILD_START";
    ChildEvents["Started"] = "CHILD_STARTED";
    ChildEvents["Restart"] = "CHILD_RESTART";
    ChildEvents["Restarted"] = "CHILD_RESTARTED";
    ChildEvents["Stop"] = "CHILD_STOP";
    ChildEvents["Stopped"] = "CHILD_STOPPED";
})(ChildEvents = exports.ChildEvents || (exports.ChildEvents = {}));
var ModulesEvents;
(function (ModulesEvents) {
    ModulesEvents["Install"] = "MODULES_INSTALL";
    ModulesEvents["Installing"] = "MODULES_INSTALLING";
    ModulesEvents["Installed"] = "MODULES_INSTALLED";
})(ModulesEvents = exports.ModulesEvents || (exports.ModulesEvents = {}));
var WatchEvents;
(function (WatchEvents) {
    WatchEvents["Start"] = "WATCH_START";
    WatchEvents["FilesChanged"] = "WATCH_FILESCHANGED";
    WatchEvents["Enable"] = "WATCH_ENABLE";
    WatchEvents["Disable"] = "WATCH_DISABLE";
    WatchEvents["Stop"] = "WATCH_STOP";
    WatchEvents["Stopped"] = "WATCH_STOPPED";
})(WatchEvents = exports.WatchEvents || (exports.WatchEvents = {}));
class EventBus extends events_1.default {
    constructor({ debug = false, } = {}) {
        super();
        this.debug = false;
        this.debug = debug;
    }
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    emit(event, ...args) {
        if (this.debug) {
            logger_1.default.prefix(`Eventbus event emit. "${event}"`);
        }
        return super.emit(event, ...args);
    }
}
exports.default = EventBus;
